egui/src/widgets/plot/mod.rs

```diff
@@ -260,50 +275,104 @@ impl Widget for Plot {
+                    if let (Some(height), Some(aspect)) = (height, view_aspect) {
+                        height * aspect
+                    } else {
+                        ui.available_size_before_wrap_finite().x
+                    }
+                })
+                .at_least(min_size.x);
+
+            let height = height
+                .unwrap_or_else(|| {
+                    if let Some(aspect) = view_aspect {
+                        width / aspect
+                    } else {
+                        ui.available_size_before_wrap_finite().y
+                    }
+                })
+                .at_least(min_size.y);
             vec2(width, height)
         };
 
         let (rect, response) = ui.allocate_exact_size(size, Sense::drag());
+        let plot_painter = ui.painter().sub_region(rect);
 
         // Background
-        ui.painter().add(Shape::Rect {
+        plot_painter.add(Shape::Rect {
             rect,
             corner_radius: 2.0,
             fill: ui.visuals().extreme_bg_color,
             stroke: ui.visuals().window_stroke(),
         });
 
+        // --- Legend ---
+
+        if show_legend {
+            // Collect the legend entries.
+            let mut legend_entries: Vec<_> = curves
+                .iter()
+                .filter(|curve| !curve.name.is_empty())
+                .map(|curve| {
+                    let checked = !hidden_curves.contains(&curve.name);
+                    let text = curve.name.clone();
+                    let color = curve.stroke.color;
+                    LegendEntry::new(text, color, checked)
+                })
+                .collect();
+
+            // Show the legend.
+            if !legend_entries.is_empty() {
+                let legend = PlotLegend::new(&mut legend_entries, rect);
+                let response = legend.ui(ui);
+                if response.hovered() {
+                    show_x = false;
+                    show_y = false;
+                }
+            }
+
+            // Get the names of the hidden curves.
+            hidden_curves = legend_entries
+                .iter()
+                .filter(|entry| !entry.checked)
+                .map(|entry| entry.text.clone())
+                .collect();
+
+            // Highlight the hovered curves.
+            legend_entries
+                .iter()
+                .filter(|entry| entry.hovered)
+                .for_each(|entry| {
+                    if let Some(curve) = curves.iter_mut().find(|curve| curve.name == entry.text) {
+                        curve.stroke.width *= 2.0;
+                    }
+                });
+
+            // Remove deselected curves.
+            curves.retain(|curve| !hidden_curves.contains(&curve.name));
+        }
+
+        // ---
+
         auto_bounds |= response.double_clicked_by(PointerButton::Primary);
 
         // Set bounds automatically based on content.
-        if auto_bounds || !bounds.is_valid() {
+        if auto_bounds {
```

Comment:
why this change? seems like good behavior to recompute bounds if they ever become invalid

Reply:
Changed it back in 6aaeae7.